### PR TITLE
Stop crashing with "Too many open files" and fixing the missing logger error

### PR DESCRIPTION
**Pull request by NeffIsBack**
## Stop crashing with "Too many open files"
With every connection we open up (if established or not) we generate **one** socket "file". These "files" are limited by the system with ulimit (you can get the soft limit value with `ulimit -n`). If there are many connections and other programs using "files" this can hit the soft limit and therefore result in the error "Too many open files".

In the PR i check if we aren't on windows (this is a UNIX thing). Than we retreive the "soft limit" and "hard limit" with `resource.getrlimit(resource.RLIMIT_NOFILE)` which returns the limit as tuple (default values: `(1024, 1048576)`). 

```python
if file_limit[1] > 10000:   # if the hard limit is above 10k
    file_limit[0] = 10000   # we set the soft limit to 10k resulting in (10000, 1048576)
```

As far as i know this only sets this limit on the current process, no system files are touched (just check with ulimit). You can also check by printing `print(resource.getrlimit(resource.RLIMIT_NOFILE))` at the start and this will always return the system default `(1024, 1048576)`.

Related PRs and resources:
- #23 
- https://github.com/Porchetta-Industries/CrackMapExec/issues/489 
- https://www.howtogeek.com/805629/too-many-open-files-linux/
- https://askubuntu.com/questions/162229/how-do-i-increase-the-open-files-limit-for-a-non-root-user
- https://stackoverflow.com/questions/56959643/what-is-pythons-equivalent-to-ulimit

## Logger issue
Fixes Issue [Porchetta #489](https://github.com/Porchetta-Industries/CrackMapExec/issues/489) where `self.logger` was not set for some reason.

Related PR: #23

### CODE REVIEW

**Comment by mpgn** [comment]
The root cause is the ulimit, increment the ulimit and the problem should be solved

Maybe we should ask to increase ulimit on the print message

**Comment by NeffIsBack** [comment]
Okay i am not that used to this ulimit stuff. If i understand that correctly we have that problem because we initiate the logger too often? If we decrease this by only ever creating one logger would that solve the issue?

**Comment by NeffIsBack** [comment]
Okay my understanding and result of my research:
Every connection (if established or not) will consume 1 "file". The maximum files allowed concurrently is defined by `ulimit -n` (kali default is 1024), therefore defines the limit on concurrent connections. So if this limit is lower, or somehow you got a lot of files open to begin with this can result in a crash when scanning large ip-ranges.

This can be increased, but should be done by the user manually, i think it would be bad practice to fiddle around with the users OS. https://askubuntu.com/questions/162229/how-do-i-increase-the-open-files-limit-for-a-non-root-user

The logger error is (from my understanding) completely independent from this. Therefore a dirty fix until we find out more

**Comment by NeffIsBack** [comment]
> Okay my understanding and result of my research: Every connection (if established or not) will consume 1 "file". The maximum files allowed concurrently is defined by `ulimit -n` (kali default is 1024), therefore defines the limit on concurrent connections. So if this limit is lower, or somehow you got a lot of files open to begin with this can result in a crash when scanning large ip-ranges.
> 
> This can be increased, but should be done by the user manually, i think it would be bad practice to fiddle around with the users OS. https://askubuntu.com/questions/162229/how-do-i-increase-the-open-files-limit-for-a-non-root-user
> 
> The logger error is (from my understanding) completely independent from this. Therefore a dirty fix until we find out more

If you want to measure the consumed files yourself put this code in line 132 (when successfully logged in):

```python
import psutil

# Funktion zum Zählen der offenen Dateien und Sockets pro Prozess mit dem Namen "python"
def count_open_files_and_sockets_per_process():
    process_counts = {}
    for proc in psutil.process_iter(['pid', 'name']):
        try:
            if proc.info['name'] == "python":
                connections = proc.connections()
                count = len(connections)
                if count > 0:
                    process_counts[proc.pid] = (proc.info['name'], count)
        except psutil.AccessDenied:
            # Ignoriere Prozesse, auf die nicht zugegriffen werden kann
            pass
    return process_counts

# Anzahl der offenen Dateien und Sockets pro Prozess mit dem Namen "python" ermitteln und filtern
process_open_files_and_sockets = count_open_files_and_sockets_per_process()
for pid, (name, count) in process_open_files_and_sockets.items():
    print(f"Prozess {name} (PID: {pid}): Anzahl der offenen Dateien und Sockets = {count}")
```

File: cme/protocols/smb.py [hunk 1]
```diff
@@ -607,7 +607,11 @@ def create_smbv3_conn(self, kdc=""):
             self.smbv1 = False
         except socket.error as e:
             if str(e).find("Too many open files") != -1:
-                self.logger.error(f"SMBv3 connection error on {self.host if not kdc else kdc}: {e}")
+                if not self.logger:
+                    print("DEBUG ERROR: logger not set, please open an issue on github: " + str(self) + str(self.logger))
```
Comment (mpgn) [inline cme/protocols/smb.py:611]: update with the ulimit workaround and we merge :)

Comment (NeffIsBack) [reply cme/protocols/smb.py:611]: I am really unsure if we should mess around with the users OS to be honest. Do you think increasing the ulimit is a good idea?

Comment (mpgn) [reply cme/protocols/smb.py:611]: It's a print, we do not change anything, the end user can decide or not to do it

**Comment by mpgn** [comment]
@lodos2005 if you want to test it :)

**Pull request merged**